lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1685,17 +1902,20 @@ class BuildControl : public FuncOpPartialLoweringPattern {
   }
 
 private:
-  /// Sequentially schedules the groups that registered themselves with
-  /// 'block'.
+  /// Schedules the groups that registered themselves with 'block'. If
+  /// sequential is true (the default for CFGs), the schedule is wrapped in a
+  /// sequential block. If sequential is false (in the case of pipelines), no
+  /// wrapper is generated.
```

Comment:
This is another place that felt a little hacky. We've been trying to share the logic here for both scf.while and staticlogic.pipeline.while. Now that the rubber is meeting the road, I've had to put more special cases for pipelines (we already had one special case, where pipelines generate a ParOp, but scf.while generates a SeqOp). I was grappling with whether to just have a separate BuildPipelineControl pattern, but decided to add a little bit more special casing (and one more parameter) to keep sharing what we can. I think in the long run this is what we want, so the insides of pipeline stages can contain things like scf.if and still get scheduled, but I'm curious if there is a better way.

Reply:
Would this not clean up by having a separate `Scheduleable` type for pipeline while? Ideally, that `Scheduleable` variant should be able to represent a hierarchy of (recursively) scheduleable "things" such that we can do as you mentioned.

Reply:
I went with this approach and removed all the isa<PipelineWhileOp> stuff in c9c8711. The one place we need to check, I added a helper `isPipelined` to WhileOpInterface.